rfcs/NNNN-lit-3.0.md

```diff
@@ -0,0 +1,112 @@
+---
+Status: Active
+Champions: @justinfagnani
+PR: https://github.com/lit/rfcs/pull/16
+---
+
+# Lit 3.0
+
+A proposal for the next major release of Lit
+
+## Objective
+
+Allow us to make a few breaking changes that will let us reduce some engineering complexity, and very lightly reduce bundle size.
+
+### Goals
+- Require no code changes from developers using non-deprecated Lit 2.0 APIs
+- Drop IE support
+- Publish ES2021
+- Small cleanups (remove Lit pre-2.0 compatibility workarounds)
+
+### Non-Goals
+- Drop support for the Shady DOM polyfill
+- Not require additional polyfills for older browsers
+
+## Motivation
+
+### IE Support removal
+
+IE11 is now unsupported by Microsoft, most major websites, many frameworks, and apparently most of our users. The times we've asked our users if they need IE support, we haven't heard from anyone that they do at all. While this doesn't mean we don't have any users who require IE11 support, all signs point to it finally being time for us to move on.
+
+Doing so will let us turn down our flaky IE11 test runners and remove some IE-specific workarounds.
+
+### Publishing ES2021
+
+We published Lit 1.0 packages as ES2019 because of spotty browser support for some ES2021 features like logical assignment. The situation is much better now, and publishing native optional chaining and logical assignment expressions will give us a small bundle size and perf improvement. 
+
+### Clean up deprecated APIs
+
+We need to start the habit of regularly removing previously deprecated APIs to keep the codebase clean.
+
+## Detailed Design
+
+### IE11 removal
+
+- Turn down our IE test runners
+- Remove workaround for IE normalizing text nodes during clone ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L974))
+- Simplify bound attribute handling ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L911))
+- Simplify `styleMap()` style removal ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/directives/style-map.ts#L82))
+- Use `globalThis` ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L15))
+- Use `toggleAttribute()` ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/reactive-element/src/reactive-element.ts#L1126))
+- Use `Element.prototype.after()`, `.before()`, `.replaceChildren()` where beneficial
```

Comment:
I measured insertBefore vs after and insertBefore is a bit faster: https://jsbench.me/wmlfm2jmh5/1 so we should stick with that.

`after()` gives a nice cleanup to `_insert()` though:

```ts
  private _insert<T extends Node>(node: T) {
    return wrap(wrap(this._$startNode).parentNode!).insertBefore(
      node,
      this._$endNode
    );
  }
```

to

```ts
  private _insert<T extends Node>(node: T) {
    wrap(this._$startNode).after(node);
    return node;
  }
```